File: ui_viewer.cpp
#include <QApplication>
#include <QWidget>
#include <QPushButton>
#include <QLabel>
#include <QMessageBox>
#include <QFile>
#include <QDebug>
#include <QScreen>
#include <QRect>
#include <QUiLoader>
#include <QVBoxLayout>

class UIViewer : public QWidget {
    Q_OBJECT
public:
    UIViewer(const QString& uiFile, QWidget *parent = nullptr) : QWidget(parent) {
        loadUI(uiFile);
        setupConnections();
        loadTheme();
    }

private slots:
    void toggleTheme() {
        currentTheme = (currentTheme == "light") ? "dark" : "light";
        loadTheme();
        
        // Обновляем текст кнопки переключения темы
        QPushButton* themeButton = findChild<QPushButton*>("themeToggle");
        if (themeButton) {
            QString text = (currentTheme == "light") ? "🌙 Темная тема" : "☀️ Светлая тема";
            themeButton->setText(text);
        }
    }

    void onButtonClicked() {
        QPushButton* button = qobject_cast<QPushButton*>(sender());
        if (button) {
            QLabel* statusLabel = findChild<QLabel*>("statusLabel");
            if (statusLabel) {
                statusLabel->setText(QString("Нажата кнопка: %1").arg(button->text()));
            }
            qDebug() << "Нажата кнопка:" << button->text();
        }
    }

private:
    QString currentTheme = "light";
    QWidget* loadedWidget = nullptr;

    void loadUI(const QString& uiFile) {
        QUiLoader loader;
        QFile file(uiFile);
        
        if (!file.open(QFile::ReadOnly)) {
            QMessageBox::critical(this, "Ошибка", 
                QString("Не удалось открыть UI файл: %1").arg(uiFile));
            return;
        }

        loadedWidget = loader.load(&file, this);
        file.close();

        if (!loadedWidget) {
            QMessageBox::critical(this, "Ошибка", 
                QString("Не удалось загрузить UI из файла: %1").arg(uiFile));
            return;
        }

        // Создаем layout для размещения загруженного виджета
        QVBoxLayout* layout = new QVBoxLayout(this);
        layout->setContentsMargins(0, 0, 0, 0);
        layout->addWidget(loadedWidget);

        // Копируем свойства окна из загруженного виджета
        setWindowTitle(loadedWidget->windowTitle());
        setMinimumSize(loadedWidget->minimumSize());
        resize(loadedWidget->size());

        qDebug() << "✅ UI файл успешно загружен!";
    }

    void setupConnections() {
        if (!loadedWidget) return;

        // Подключаем кнопку переключения темы
        QPushButton* themeButton = loadedWidget->findChild<QPushButton*>("themeToggle");
        if (themeButton) {
            connect(themeButton, &QPushButton::clicked, this, &UIViewer::toggleTheme);
        }

        // Подключаем все остальные кнопки
        QList<QPushButton*> buttons = loadedWidget->findChildren<QPushButton*>();
        for (QPushButton* button : buttons) {
            if (button != themeButton) {
                connect(button, &QPushButton::clicked, this, &UIViewer::onButtonClicked);
            }
        }
    }

    void loadTheme() {
        QString fileName = QString("%1_theme.qss").arg(currentTheme);
        QFile file(fileName);
        
        if (file.open(QFile::ReadOnly)) {
            QString styleSheet = QString::fromLatin1(file.readAll());
            qApp->setStyleSheet(styleSheet);
            qDebug() << "🎨 Тема" << currentTheme << "загружена";
        } else {
            qDebug() << "⚠️ Не удалось загрузить тему" << fileName;
        }
    }
};

int main(int argc, char *argv[])
{
    QApplication app(argc, argv);
    
    // Устанавливаем свойства приложения
    app.setApplicationName("UI Viewer");
    app.setApplicationVersion("1.0");
    app.setOrganizationName("Demo Company");

    QString uiFile = "demo.ui";
    
    // Проверяем существование UI файла
    if (!QFile::exists(uiFile)) {
        QMessageBox::critical(nullptr, "Ошибка", 
            QString("UI файл не найден: %1\n\nУбедитесь, что файл находится в текущей директории").arg(uiFile));
        return 1;
    }

    // Проверяем файлы стилей
    QStringList styleFiles = {"light_theme.qss", "dark_theme.qss"};
    QStringList missingFiles;
    
    for (const QString& file : styleFiles) {
        if (!QFile::exists(file)) {
            missingFiles << file;
        }
    }
    
    if (!missingFiles.isEmpty()) {
        QMessageBox::warning(nullptr, "Предупреждение", 
            QString("Не найдены файлы стилей: %1\n\nПриложение будет работать без стилей").arg(missingFiles.join(", ")));
    }

    // Создаем и показываем окно
    UIViewer viewer(uiFile);
    viewer.show();

    // Центрируем окно на экране
    QScreen* screen = app.primaryScreen();
    QRect screenGeometry = screen->availableGeometry();
    QRect windowGeometry = viewer.frameGeometry();
    QPoint centerPoint = screenGeometry.center();
    windowGeometry.moveCenter(centerPoint);
    viewer.move(windowGeometry.topLeft());

    qDebug() << "🚀 UI Viewer запущен!";
    qDebug() << "💡 Используйте кнопку переключения темы для смены стилей";

    return app.exec();
}

#include "ui_viewer.moc"

File: ui_viewer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Простой просмотрщик UI файлов Qt без Designer
"""

import sys
import os
from PyQt5 import QtWidgets, uic

def load_stylesheet(path):
    """Загрузить QSS стили из файла"""
    try:
        with open(path, 'r', encoding='utf-8') as file:
            return file.read()
    except Exception as e:
        print(f"Не удалось загрузить стили {path}:", e)
        return ""

class UIViewer(QtWidgets.QMainWindow):
    def __init__(self, ui_file):
        super().__init__()
        self.current_theme = "light"
        
        # Загружаем UI файл прямо в self (поскольку demo.ui создает QMainWindow)
        try:
            uic.loadUi(ui_file, self)
            print(f"✅ UI файл {ui_file} успешно загружен!")
        except Exception as e:
            print(f"❌ Ошибка загрузки UI файла {ui_file}: {e}")
            sys.exit(1)
        
        self.setup_connections()
        self.load_theme()
        
    def setup_connections(self):
        """Настройка соединений сигналов"""
        # Подключаем кнопку переключения темы
        if hasattr(self, 'themeButton'):
            self.themeButton.clicked.connect(self.toggle_theme)
        
        # Подключаем все остальные кнопки к общему обработчику
        buttons = self.findChildren(QtWidgets.QPushButton)
        for button in buttons:
            if button != getattr(self, 'themeButton', None):
                button.clicked.connect(lambda checked, btn=button: self.on_button_clicked(btn))
    
    def toggle_theme(self):
        """Переключение темы"""
        self.current_theme = "dark" if self.current_theme == "light" else "light"
        self.load_theme()
        
        # Обновляем текст кнопки
        if hasattr(self, 'themeButton'):
            if self.current_theme == "light":
                self.themeButton.setText("Dark theme")
            else:
                self.themeButton.setText("Light theme")
    
    def load_theme(self):
        """Загрузка текущей темы"""
        theme_file = f"{self.current_theme}_theme.qss"
        style = load_stylesheet(theme_file)
        
        if style:
            QtWidgets.QApplication.instance().setStyleSheet(style)
            print(f"🎨 Тема {self.current_theme} загружена")
        else:
            print(f"⚠️ Не удалось загрузить тему {theme_file}")
    
    def on_button_clicked(self, button):
        """Обработчик нажатия кнопок"""
        print(f"🔘 Нажата кнопка: {button.text()}")

def main():
    """Главная функция"""
    app = QtWidgets.QApplication(sys.argv)
    
    # Проверяем наличие UI файла
    ui_file = "demo.ui"
    if not os.path.exists(ui_file):
        print(f"❌ UI файл {ui_file} не найден!")
        print("📍 Убедитесь, что файл находится в текущей директории")
        return
    
    # Проверяем наличие файлов стилей
    required_files = ["light_theme.qss", "dark_theme.qss"]
    missing_files = [f for f in required_files if not os.path.exists(f)]
    
    if missing_files:
        print(f"⚠️ Не найдены файлы стилей: {', '.join(missing_files)}")
        print("🎨 Приложение будет работать без стилей")
    
    # Создаем и показываем окно
    try:
        viewer = UIViewer(ui_file)
        viewer.show()
        
        # Центрируем окно на экране
        screen = app.primaryScreen()
        screen_geometry = screen.availableGeometry()
        window_geometry = viewer.frameGeometry()
        center_point = screen_geometry.center()
        window_geometry.moveCenter(center_point)
        viewer.move(window_geometry.topLeft())
        
        print("🚀 UI Viewer запущен!")
        print("💡 Используйте кнопку переключения темы для смены стилей")
        
        sys.exit(app.exec_())
        
    except Exception as e:
        print(f"❌ Ошибка запуска приложения: {e}")
        return

if __name__ == "__main__":
    main()
